src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexWriter.cs

```diff
@@ -18,89 +18,118 @@
 
 namespace System.Text.RegularExpressions
 {
-    internal sealed class RegexWriter
+    internal ref struct RegexWriter
     {
-        private int[] _intStack;
-        private int _depth;
-        private int[] _emitted;
-        private int _curpos;
-        private readonly Dictionary<string, int> _stringhash;
-        private readonly List<string> _stringtable;
-        private bool _counting;
-        private int _count;
-        private int _trackcount;
-        private Hashtable _caps;
-
         private const int BeforeChild = 64;
         private const int AfterChild = 128;
 
-        /// <summary>
-        /// This is the only function that should be called from outside.
-        /// It takes a RegexTree and creates a corresponding RegexCode.
-        /// </summary>
-        internal static RegexCode Write(RegexTree t)
-        {
-            RegexWriter w = new RegexWriter();
-            RegexCode retval = w.RegexCodeFromRegexTree(t);
-#if DEBUG
-            if (t.Debug)
-            {
-                t.Dump();
-                retval.Dump();
-            }
-#endif
-            return retval;
-        }
+        public ValueListBuilder<int> Emitted;
+        public ValueListBuilder<int> IntStack;
+        public readonly Dictionary<string, int> StringHash;
+        public readonly List<string> StringTable;
+        public Hashtable Caps;
+        public int TrackCount { get; private set; }
 
-        // Private constructor; can't be created outside
-        private RegexWriter()
+        private RegexWriter(Span<int> emittedSpan, Span<int> intStackSpan)
         {
-            _intStack = new int[32];
-            _emitted = new int[32];
-            _stringhash = new Dictionary<string, int>();
-            _stringtable = new List<string>();
+            Emitted = new ValueListBuilder<int>(emittedSpan);
+            IntStack = new ValueListBuilder<int>(intStackSpan);
+            StringHash = new Dictionary<string, int>();
+            StringTable = new List<string>();
+            Caps = null;
+            TrackCount = 0;
         }
 
         /// <summary>
-        /// To avoid recursion, we use a simple integer stack.
-        /// This is the push.
+        /// The top level RegexCode generator. It does a depth-first walk
+        /// through the tree and calls EmitFragment to emits code before
+        /// and after each child of an interior node, and at each leaf.
         /// </summary>
-        private void PushInt(int i)
+        public static RegexCode Write(RegexTree tree)
         {
-            if (_depth >= _intStack.Length)
+            Span<int> emittedSpan = stackalloc int[48];
+            Span<int> intStackSpan = stackalloc int[32];
+            RegexWriter writer = new RegexWriter(emittedSpan, intStackSpan);
+
+            // construct sparse capnum mapping if some numbers are unused
+            int capsize;
+            if (tree._capnumlist == null || tree._captop == tree._capnumlist.Length)
```

Comment:
Why did you manually inline RegexCodeFromRegexTree?  It makes it difficult to see any differences.  Was it necessary for some reason?

Reply:
It wasn't really necessary, that might fall under over-engineering. I can switch back to what it was before.